File: dx-api/app/helpers/scoring.go
package helpers

import "dx-api/app/consts"

// ComboState tracks the player's combo streak within a level.
type ComboState struct {
	Streak        int
	CyclePosition int
	TotalScore    int
	MaxCombo      int
}

// ProcessAnswerResult holds the scoring outcome of a single answer.
type ProcessAnswerResult struct {
	State       ComboState
	PointsEarned int
	ComboBonus   int
}

// ProcessAnswer computes scoring for a single answer based on current combo state.
func ProcessAnswer(state ComboState, isCorrect bool) ProcessAnswerResult {
	if !isCorrect {
		return ProcessAnswerResult{
			State: ComboState{
				Streak:        0,
				CyclePosition: 0,
				TotalScore:    state.TotalScore,
				MaxCombo:      state.MaxCombo,
			},
		}
	}

	points := consts.CorrectAnswer
	bonus := 0
	newStreak := state.Streak + 1
	newCyclePosition := state.CyclePosition + 1

	for _, threshold := range consts.ComboThresholds {
		if newCyclePosition == threshold.Streak {
			bonus += threshold.Bonus
		}
	}

	if newCyclePosition >= consts.ComboCycleLength {
		newCyclePosition = 0
	}

	points += bonus

	newMaxCombo := state.MaxCombo
	if newStreak > newMaxCombo {
		newMaxCombo = newStreak
	}

	return ProcessAnswerResult{
		State: ComboState{
			Streak:        newStreak,
			CyclePosition: newCyclePosition,
			TotalScore:    state.TotalScore + points,
			MaxCombo:      newMaxCombo,
		},
		PointsEarned: points,
		ComboBonus:   bonus,
	}
}

File: dx-web/src/features/web/play-core/helpers/scoring.ts
import { SCORING, COMBO_THRESHOLDS } from "@/consts/scoring";

export type ComboState = {
  streak: number;
  cyclePosition: number;
  totalScore: number;
  maxCombo: number;
};

export function createComboState(): ComboState {
  return { streak: 0, cyclePosition: 0, totalScore: 0, maxCombo: 0 };
}

export function processAnswer(
  state: ComboState,
  isCorrect: boolean
): { state: ComboState; pointsEarned: number; comboBonus: number } {
  if (!isCorrect) {
    return {
      state: { ...state, streak: 0, cyclePosition: 0 },
      pointsEarned: 0,
      comboBonus: 0,
    };
  }

  let points = SCORING.CORRECT_ANSWER;
  let bonus = 0;
  const newStreak = state.streak + 1;
  let newCyclePosition = state.cyclePosition + 1;

  for (const threshold of COMBO_THRESHOLDS) {
    if (newCyclePosition === threshold.streak) {
      bonus += threshold.bonus;
    }
  }

  if (newCyclePosition >= SCORING.COMBO_CYCLE_LENGTH) {
    newCyclePosition = 0;
  }

  points += bonus;

  return {
    state: {
      streak: newStreak,
      cyclePosition: newCyclePosition,
      totalScore: state.totalScore + points,
      maxCombo: Math.max(state.maxCombo, newStreak),
    },
    pointsEarned: points,
    comboBonus: bonus,
  };
}
